library/std/src/thread/local.rs

```diff
@@ -1137,12 +1137,20 @@ pub mod os {
         //
         // Note that to prevent an infinite loop we reset it back to null right
         // before we return from the destructor ourselves.
-        unsafe {
+        //
+        // The return value of `panic::catch_unwind` must not be dropped before
+        // `always_abort` is called, as the drop function of the value can panic
+        // itself, causing an unwinding to happen outside the `catch_unwind`
+        // call.
+        if let Err(_err) = crate::panic::catch_unwind(|| unsafe {
             let ptr = Box::from_raw(ptr as *mut Value<T>);
             let key = ptr.key;
             key.os.set(ptr::invalid_mut(1));
             drop(ptr);
             key.os.set(ptr::null_mut());
+        }) {
+            crate::panic::always_abort();
+            panic!("destruction of a thread local value");
```

Comment:
slightly more descriptive: "panic during destruction..."

Reply:
The two lines have been replaced by `rtabort!("destructor panicked");` -- is this okay?